### PR TITLE
Enhance dynamic hood control for trench clearance

### LINKED ISSUE #47: [FEATURE] Zone-based control of hood
### Description:
On the V2 robot, the shooter will be lowered, and the hood may or may not extend beyond the height of the trench. If it does, look into adding a zone in which the hood will automatically clear itself of the trench. Consider limiting the maximum desired hood angle based on this zone.

#### Type:
- [ ] New Feature
- [x] Improvement to Existing Feature
### Category:
- [ ] Software
- [x] Controls/Mechanism
- [ ] Controllers/Buttons/Human Interface
- [ ] UI/Dashboard
- [ ] Logging

---
### Urgency
- [ ] Nice to Have
- [x] Important
- [ ] Critical
### Impacted Modes:
- [ ] Disabled
- [x] Teleop
- [x] Auto
- [ ] Simulation
### Impacted Systems:
- [ ] Drive
- [ ] Vision
- [ ] Autonomous
- [ ] Logging
- [x] Mechanism(s): Hood
### When?
- [ ] Periodic
- [ ] Init
- [x] Command
### File(s) of Focus:
`HoodSystem.java`
`FlywheelSystem.java`
`PositionDetails.java`
`positionDetails.json`
### Resources:

---

### PR DESCRIPTION
**Pull request by skrunked**
Updates the positionDetails JSON and parser to have trench zones, underneath which the hood will be set to its maximum angle to avoid clipping the trench. Also includes _many_ additional comments in PositionDetails.java and other miscellaneous refactoring. 
Closes #47.

https://github.com/user-attachments/assets/ae47e474-c92d-4349-afaf-b9c1f89fc915

- [x] Tested on Robot
- [x] Tested in Sim

### CODE REVIEW

**Comment by skrunked** [comment]
Actually, should we make a new branch to re-export the AS model and verify the turret offset before we send this?

**Comment by skrunked** [comment]
More things while I'm thinking about it:
- make zoning based on turret position rather than robot center (add offset to contains() method of Zone?)
- add drivebase speed modifier when in trench zone
- evaluate after testing on real robot, but potentially increase size of trench zones horizontally to give enough buffer when going under trench at speed (real hood is more responsive than sim hood, so might be a non-issue)

**Comment by rxgran** [comment]
> More things while I'm thinking about it:
> 
> * make zoning based on turret position rather than robot center (add offset to contains() method of Zone?)
> * add drivebase speed modifier when in trench zone
> * evaluate after testing on real robot, but potentially increase size of trench zones horizontally to give enough buffer when going under trench at speed (real hood is more responsive than sim hood, so might be a non-issue)

I don't think it would make a major difference re-doing the zoning based on turret position and it might be slightly un-intuitive to the driver if we do that at edge cases. 

Speed modifier might be nice

Definitely test on robot before merge.

File: src/main/java/org/team157/robot/parsing/PositionDetails.java [hunk 1]
```diff
@@ -16,114 +16,261 @@
...
+ */
 public class PositionDetails {
+    /** Location of the JSON file containting position information, placed inside the deploy directory. */
     private final String jsonPath = "/positionDetails.json";
+
+    /** Location representing the center of the blue alliance's Hub. */
+    public Location blueHubLocation;
+    /** Location representing the center of the red alliance's Hub. */
+    public Location redHubLocation;    
+    // TODO: Consider changing names of passing locations to not include the word 'zone' to avoid confusion with the actual passing zones defined as rectangles.
```
Comment (rxgran) [inline src/main/java/org/team157/robot/parsing/PositionDetails.java:34]: maybe `<xxx>PassingTarget<xxx>`

File: src/main/deploy/positionDetails.json [hunk 2]
```diff
@@ -57,6 +57,30 @@
             "yMin": 4.040981,
             "xMax": 16.535400,
             "yMax": 7.98703
+        },
+        "blueTrenchLow": {
```
Comment (rxgran) [inline src/main/deploy/positionDetails.json:61]: Consider making these larger or smaller after testing on robot

File: src/main/java/org/team157/robot/subsystems/drive/DriveSystem.java [hunk 3]
```diff
@@ -369,4 +371,12 @@ public Optional<Pose2d> samplePoseAt(double timestampSeconds) {
     public Pose2d getPose() {
         return super.getStateCopy().Pose;
     }
+    /**
+     * Returns whether or not the turret is underneath the trench.
+     * 
+     * @return true if the turret (offset from drivebase center) is within a trench zone, false otherwise.
+     */
+    public boolean isUnderTrench() {
+        return FieldConstants.positionDetails.isUnderTrench(getPose().plus(ModelConstants.XY_ORIGIN_TO_TURRET_BASE_OFFSET));
```
Comment (rxgran) [inline src/main/java/org/team157/robot/subsystems/drive/DriveSystem.java:381]: like this is probably ok but it might be un-intuitive to the driver if we are on the edge of the line and the zone is caring about the turret rather than the robot's position. Although it might be safer if the turret is leading toward the trench... your call

File: src/main/java/org/team157/robot/subsystems/flywheel/FlywheelSystem.java [hunk 4]
```diff
@@ -108,8 +109,16 @@ static double velocityFunction(double distance, double height, double theta) {
     return Math.sqrt((g * distance * distance) / denominator);
   }
 
+  double getLowerBound(boolean isUnderTrench){
+    if(isUnderTrench){
+      return HoodConstants.UPPER_SOFT_LIMIT.in(Radians);
```
Comment (rxgran) [inline src/main/java/org/team157/robot/subsystems/flywheel/FlywheelSystem.java:114]: we might have a little wiggle room to play with here, find out in testing with robot.

**Review by rxgran — CHANGES_REQUESTED** [review]
just some conceptual comments. Will definitely need to be tested on the robot first before merging.
